llpc/context/llpcCompiler.cpp

```diff
@@ -1418,12 +1429,162 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+    llvm::Optional<CacheAccessor> cacheAccessor;
+    cacheAccessor.emplace(context, partPipelineHash, getInternalCaches());
+    if (cacheAccessor->isInCache()) {
+      // We got a cache hit.
+      if (partIdx == 0)
+        LLPC_OUTS("Fragment shader cache hit.\n");
+      else
+        LLPC_OUTS("Non fragment shader cache hit.\n");
+      // Mark the applicable entries in stageCacheAccesses.
+      for (unsigned shaderStage = 1 << llvm::countTrailingZeros(partStageMask); 1 << shaderStage <= partStageMask;
+           ++shaderStage) {
+        if (partStageMask & (1 << shaderStage)) {
+          stageCacheAccesses[shaderStage] =
+              cacheAccessor->hitInternalCache() ? CacheAccessInfo::InternalCacheHit : CacheAccessInfo::CacheHit;
+        }
+      }
+      // Get the ELF from the cache.
+      partPipelineElf = llvm::StringRef(static_cast<const char *>(cacheAccessor->getElfFromCache().pCode),
+                                        cacheAccessor->getElfFromCache().codeSize);
+    } else {
+      // Cache miss.
+      if (partIdx == 0)
+        LLPC_OUTS("Fragment shader cache miss.\n");
+      else
+        LLPC_OUTS("Non fragment shader cache miss.\n");
+      // Compile the part pipeline.
+      // Keep the PipelineShaderInfo structs for the shaders we are including.
+      llvm::SmallVector<const PipelineShaderInfo *, 4> partPipelineShaderInfo;
+      for (const PipelineShaderInfo *oneShaderInfo : shaderInfo) {
+        if (oneShaderInfo && (partStageMask & (1 << oneShaderInfo->entryStage)))
+          partPipelineShaderInfo.push_back(oneShaderInfo);
+      }
+      result = buildPipelineInternal(context, partPipelineShaderInfo, PipelineLink::PartPipeline, &*elfLinkerPipeline,
+                                     &partPipelineBuffers[partIdx], stageCacheAccesses);
+      if (result != Result::Success)
+        break;
+      partPipelineElf = partPipelineBuffers[partIdx];
+
+      // If the "ELF" does not look like ELF, then it must be textual output from -emit-lgc, -emit-llvm, -filetype=asm.
+      // We can't link that, so just concatenate it on to the output.
+      if (partPipelineElf.size() < 4 || !partPipelineElf.startswith("\177ELF")) {
+        *pipelineElf += partPipelineElf;
+        continue;
+      }
+
+      // Store the part-pipeline ELF in the cache.
+      cacheAccessor->setElfInCache(BinaryData({partPipelineElf.size(), partPipelineElf.data()}));
+    }
+
+    if (Llpc::EnableOuts()) {                                                                                          \
+      ElfReader<Elf64> reader(m_gfxIp);
+      size_t readSize = 0;
+      if (reader.ReadFromBuffer(partPipelineElf.data(), &readSize) == Result::Success) {
+        LLPC_OUTS("===============================================================================\n");
+        LLPC_OUTS("// LLPC part-pipeline ELF (from cache or just compiled)\n");
+        LLPC_OUTS(reader);
+      }
+    }
+
+    if (!elfLinker) {
```

Comment:
The code in this block does not need to be in the loop.  I get that it does not have to be executed if the compilation of the fragment part fails, but I expect that to be so rare that it is not worth cluttering the loop.  As I mentioned above, I would remove the use of the unique_ptr, and just make elfLinkerPipeline and elfLinker regular stack objects.  Then all of this could be part of the initialization of these objects at the start of the function.

Reply:
Moved this up.